ert/data/_record.py

```diff
@@ -131,34 +135,59 @@ class BlobRecord(Record):
+class RecordCollectionType(str, Enum):
+    NON_UNIFORM = "NON_UNIFORM"
+    UNIFORM = "UNIFORM"
+
+
+_RecordTupleType = Union[Tuple[NumericalRecord, ...], Tuple[BlobRecord, ...]]
+
+
+class RecordCollection:
+    def __init__(
+        self,
+        records: Tuple[Record, ...],
+        ensemble_size: Optional[PositiveInt] = None,
+        collection_type: RecordCollectionType = RecordCollectionType.NON_UNIFORM,
+    ):
+        if len(records) < 1:
+            raise ValueError("At least one record must be provided")
+        if collection_type == RecordCollectionType.UNIFORM:
+            if len(records) > 1:
+                raise ValueError("Multiple records provided for a uniform record")
+            if ensemble_size is None:
+                raise ValueError("Ensemble size missing for uniform record")
+            self._records = cast(_RecordTupleType, records * ensemble_size)
+            self._ensemble_size = ensemble_size
+        else:
+            if ensemble_size is not None and ensemble_size != len(records):
+                raise ValueError("Ensemble size does not match the record count")
+            for record in records:
+                if record.record_type != records[0].record_type:
+                    raise ValueError("Ensemble records must have a uniform record type")
+            self._records = cast(_RecordTupleType, records)
+            self._ensemble_size = len(self._records)
+        self._collection_type = collection_type
+
+    def __eq__(self, other: object) -> bool:
+        return self.__dict__ == other.__dict__
 
     @property
-    def record_type(self) -> Optional[RecordType]:
-        return self.records[0].record_type
-
-    @validator("ensemble_size", pre=True, always=True)
-    def ensemble_size_validator(
-        cls, ensemble_size: Optional[int], values: Dict[str, Any]
-    ) -> Optional[int]:
-        if ensemble_size is None and "records" in values:
-            ensemble_size = len(values["records"])
-        assert ensemble_size is not None and ensemble_size > 0
-        return ensemble_size
+    def records(self) -> _RecordTupleType:
+        return self._records
 
-    @root_validator(skip_on_failure=True)
-    def ensure_consistent_ensemble(
-        cls, ensemble_record: Dict[str, Any]
-    ) -> Dict[str, Any]:
-        assert "records" in ensemble_record and "ensemble_size" in ensemble_record
-        assert len(ensemble_record["records"]) == ensemble_record["ensemble_size"]
-        record_type = ensemble_record["records"][0].record_type
-        for record in ensemble_record["records"][1:]:
-            if record.record_type != record_type:
-                raise ValueError("Ensemble records must have a uniform record type")
-        return ensemble_record
+    @property
+    def ensemble_size(self) -> int:
+        return self._ensemble_size
+
+    @property
+    def record_type(self) -> RecordType:
+        assert self._records[0].record_type is not None  # mypy needs this
```

Comment:
Yeah because `record_type` is `Optional`. Ouch.

Thanks for the comment!

Reply:
That is a consequence of the way we used `pydantic`, we just made that optional, so that we would could pass `None` to indicate that it should be derived. Then you end up with a data record that according to the static typing could have a `None` type... That is one of those things that make `pydantic` a bit limited in my view (great for some things though).